zkvm/src/types.rs

```diff
@@ -344,6 +420,31 @@ impl Contract {
     }
 }
 
+impl FrozenContract {
+    fn encode(&self, program: &mut Vec<u8>) {
+        program.extend_from_slice(&self.predicate.to_point().to_bytes());
+        for p in self.payload.iter() {
+            match p {
+                // Data = 0x00 || LE32(len) || <bytes>
+                FrozenItem::Data(d) => {
+                    program.push(0u8);
+                    let bytes = d.to_bytes();
+                    // If we can't get the length before serializing, might have
+                    // to write a filler value and then go backwards
```

Comment:
this extra allocation on the prover side probably isn't going to matter anyway, so this is good enough, imho